### PR TITLE
Optimizing RNAseq Elasticnet models

### PR DESCRIPTION
**Pull request by jaceybronte**
This PR adds the Elasticnet models for incorporating RNAseq data. There is one optimized model for every single latent dimension found by BioBombe.

### CODE REVIEW

File: 6.RNAseq/scripts/3.models.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+# In[1]:
+
+
+import pandas as pd
+import pathlib
+import sys
+from sklearn.preprocessing import MinMaxScaler
+import joblib
+import torch
+from torch.utils.data import DataLoader, TensorDataset
```
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py]: consider running a pre-commit for isort here

File: 6.RNAseq/scripts/3.models.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+from model_utils import extract_latent_dims
+
+sys.path.insert(0,"./utils")
+from optimizing_utils import model_training
+
+
+# In[2]:
+
+
+def align_dataframes(train, test, val, latent_df):
```
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py]: consider adding type hints and/or being slightly more descriptive about your var names. I like how latent_df had information in the name that it is a df

File: 6.RNAseq/scripts/3.models.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    - latent_df (pd.DataFrame): Dataframe containing latent dimensions with `ModelID` as the key.
+
+    Returns:
+    - X_train_features (pd.DataFrame): Training dataset features (gene expression).
+    - X_val_features (pd.DataFrame): Validation dataset features (gene expression).
+    - X_test_features (pd.DataFrame): Combined validation & test dataset features (gene expression).
+    - y_train (pd.DataFrame): Training dataset target (latent dimensions).
+    - y_val (pd.DataFrame): Validation dataset target (latent dimensions).
+    - y_test (pd.DataFrame): Combined validation & test dataset target (latent dimensions).
+    """
```
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py:53]: optionally return a tuple of vars

File: 6.RNAseq/scripts/3.models.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    y_val = val[latent_columns]
+    y_test = test[latent_columns]
+
+    # Identify zero-only columns in y_train
+    zero_columns = y_train.columns[(y_train == 0).all()]
+
+    # Drop these columns from all y datasets
+    y_train = y_train.drop(columns=zero_columns)
+    y_val = y_val.drop(columns=zero_columns)
+    y_test = y_test.drop(columns=zero_columns)
```
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py:78]: Why? What do these columns represent?

File: 6.RNAseq/scripts/3.models.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+train_and_test_subbed_dir = pathlib.Path("../0.data-download/data/train_and_test_subbed.parquet").resolve()
+train_and_test_subbed = pd.read_parquet(train_and_test_subbed_dir)
+
+# Convert DataFrame to NumPy and then Tensor
+train_test_array = train_and_test_subbed.to_numpy()
+train_test_tensor = torch.tensor(train_test_array, dtype=torch.float32)
+
+#Create TensorDataset and DataLoader
+tensor_dataset = TensorDataset(train_test_tensor)
+train_and_test_subbed_loader = DataLoader(tensor_dataset, batch_size=32, shuffle=False)
```
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py:136]: for these dataloaders, was batch size optimized? Also why or why not shuffle the order here? I might answer that in later code.

File: 6.RNAseq/utils/optimizing_utils.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import joblib
+import optuna
+import pandas as pd
+from joblib import dump
+from sklearn.linear_model import ElasticNet
+from sklearn.metrics import r2_score, mean_squared_error
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import StandardScaler
+
+def optimize_elasticnet(trial, X_train, y_train, X_val, y_val):
```
Comment (MikeLippincott) [inline 6.RNAseq/utils/optimizing_utils.py]: same general function comments here for type hints, returns, etc

File: 6.RNAseq/utils/optimizing_utils.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    l1_ratio = trial.suggest_float('l1_ratio', 0.0, 1.0)
+    
+    # Build and train the ElasticNet model
+    model = make_pipeline(StandardScaler(), ElasticNet(alpha=alpha, l1_ratio=l1_ratio, max_iter=10000))
+    model.fit(X_train, y_train)
+    
+    # Evaluate on validation data
+    y_pred = model.predict(X_val)
+    r2 = r2_score(y_val, y_pred)
+    return r2
```
Comment (MikeLippincott) [inline 6.RNAseq/utils/optimizing_utils.py:41]: Question: is R2 score the best loss function to optimize hyperparameters for?

File: 6.RNAseq/scripts/3.models.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+for model_file in model_save_dir.glob("*.joblib"):
+    model_file_name = model_file.stem
+    try:
+        parts = model_file_name.split("_")
+        model_name = parts[0]
+        num_components = int(parts[3])
+        init = int(parts[7])
+        seed = int(parts[9])
+    except (IndexError, ValueError):
+        print(f"Skipping file {model_file} due to unexpected filename format.")
```
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py:184]: For these tpyes of edge case catches throughout this PR is there a log or file that is being written to show unprocessed models? How do we ensure that some are not completely overlooked or forgotten?

File: 6.RNAseq/scripts/3.models.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    # Drop these columns from all y datasets
+    y_train = y_train.drop(columns=zero_columns)
+    y_val = y_val.drop(columns=zero_columns)
+    y_test = y_test.drop(columns=zero_columns)
+
+    X_test_features = pd.concat([X_val_features, X_test_features], axis=0)
+    y_test = pd.concat([y_val, y_test], axis=0)
+
+    return X_train_features, X_val_features, X_test_features, y_train, y_val, y_test
+
```
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py]: Why combine the val and test datasets here?

File: 6.RNAseq/scripts/3.models.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    # Filter latent_df to only include missing z_dimensions (ensure the column names are integers)
+    latent_df = latent_df[[col for col in latent_df.columns if (isinstance(col, int) and col in missing_z_dimensions) or col == 'ModelID']]
+
+    # Align dataframes for training and testing
+    x_train, x_val, x_test, y_train, y_val, y_test = align_dataframes(train, test, val, latent_df)
+
+    # Perform model training and prediction only for missing z_dimensions
+    test_results_df, test_predictions_df = model_training(
+        x_train, x_val, x_test, y_train, y_val, y_test, model_name, num_components, init, seed
+    )
```
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py:233]: This essentially leaks the val data into the testing data yes? The val data was used to pick hyperparameters though? Unless I am not understanding a key component of the design - I would avoid such leakage.

Please let me know if I am missing a key part though!

File: 6.RNAseq/scripts/3.models.py
Comment (MikeLippincott) [inline 6.RNAseq/scripts/3.models.py:1]: Consider adding shuffled models to assess if certain z-dimensions are predicted better due to feature space size or distibutions

**Review by MikeLippincott — APPROVED** [review]
I noted a potential data leak in the model assessment process. There are some minor documentation concerns for functions and I would consider adding a README explaining the goals of the model and how many models are ultimately being trained.

File: 6.RNAseq/scripts/3.models.py (hunk 10 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py:233]: Oh excellent catch, this was an artifact from the original RNAseq data where we needed to include val and test to have more than 5 data points haha

File: 6.RNAseq/scripts/3.models.py (hunk 9 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py]: artifact!

File: 6.RNAseq/scripts/3.models.py (hunk 8 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py:184]: Yep it's logged, though this should never occur unless there are errors earlier that need to be noted

File: 6.RNAseq/scripts/3.models.py (hunk 4 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py:78]: A column with entirely 0s would represent a multi gene process that is entirely irrelevant in every single sample. There are no tumors in which that process has any effect, positive or negative

File: 6.RNAseq/scripts/3.models.py (hunk 1 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py]: done

File: 6.RNAseq/scripts/3.models.py (hunk 2 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py]: done

File: 6.RNAseq/scripts/3.models.py (hunk 5 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py:136]: This is just for efficiency - batchsize doesn't necessarily need to be perfect, it is just loading the data in a separate function one batch at a time. Shuffled is just an argument for the DataLoader, and it isn't necessary to shuffle here

File: 6.RNAseq/utils/optimizing_utils.py (hunk 6 above)
Comment (jaceybronte) [reply 6.RNAseq/utils/optimizing_utils.py]: done

File: 6.RNAseq/scripts/3.models.py
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py:1]: This is a good idea!

File: 6.RNAseq/scripts/3.models.py (hunk 1 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py]: done

File: 6.RNAseq/scripts/3.models.py (hunk 2 above)
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py]: done

File: 6.RNAseq/scripts/3.models.py
Comment (jaceybronte) [reply 6.RNAseq/scripts/3.models.py:1]: opened an issue

File: 6.RNAseq/utils/optimizing_utils.py (hunk 7 above)
Comment (jaceybronte) [reply 6.RNAseq/utils/optimizing_utils.py:41]: opened an issue
